services/ens/ensresolver/resolver.go

```diff
@@ -83,20 +84,57 @@ func (e *EnsResolver) OwnerOf(ctx context.Context, chainID uint64, username stri
 		return nil, err
 	}
 
+	nameHash := walletCommon.NameHash(username)
+
 	registry, err := e.contractMaker.NewRegistry(chainID)
 	if err != nil {
 		return nil, err
 	}
 
 	callOpts := &bind.CallOpts{Context: ctx, Pending: false}
-	owner, err := registry.Owner(callOpts, walletCommon.NameHash(username))
+	owner, err := registry.Owner(callOpts, nameHash)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the NameWrapper contract address for the given chain ID
+	nameWrapperAddress, err := walletCommon.GetNameWrapperAddress(chainID)
 	if err != nil {
 		return nil, err
 	}
 
+	// If the owner is the NameWrapper contract, resolve the true owner
+	if owner == nameWrapperAddress {
+		owner, err = e.resolveWrappedOwner(ctx, chainID, nameHash, nameWrapperAddress)
+		if err != nil {
+			return nil, err
+		}
+	}
+
 	return &owner, nil
 }
 
+// resolveWrappedOwner resolves the actual ENS owner from the NameWrapper contract
+// for wrapped names (those owned by the wrapper itself in the ENS registry).
+func (e *EnsResolver) resolveWrappedOwner(ctx context.Context, chainID uint64, nameHash common.Hash, wrapperAddr common.Address) (common.Address, error) {
+	callOpts := &bind.CallOpts{Context: ctx, Pending: false}
+
+	nameWrapper, err := e.contractMaker.NewNameWrapper(chainID, &wrapperAddr)
+	if err != nil {
+		return common.Address{}, err
```

Comment:
I'd generally prefer using `*common.Address` as return value and returning nil in all error conditions. The empty address (0x0) is a valid address after all :D


Moreover, it lets you use it directly to return from the function calling this one.